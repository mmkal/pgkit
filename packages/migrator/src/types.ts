import {sql, Client, Queryable} from '@pgkit/client'
import {type MigraOptions} from '@pgkit/migra'
import * as umzug from 'umzug'

export interface MigratorContext {
  connection: Queryable
  sql: typeof sql
}

export type Awaitable<T> = T | Promise<T>

export type Migration = (params: umzug.MigrationParams<MigratorContext>) => Promise<void>

export type Confirm = (sql: string, options?: {readonly?: boolean}) => Awaitable<string | null>

export interface BaseListedMigration {
  name: string
  path: string
  content: string
  note?: string
}

export interface PendingMigration extends BaseListedMigration {
  status: 'pending'
}

export interface ExecutedMigration extends BaseListedMigration {
  status: 'executed'
  drifted: boolean
}

export type ListedMigration = PendingMigration | ExecutedMigration

export type RunnableMigration = umzug.RunnableMigration<MigratorContext>

export type Logger = {
  info: (...args: unknown[]) => void
  warn: (...args: unknown[]) => void
  error: (...args: unknown[]) => void
}

export type Task = <T>(name: string, fn: () => Promise<T>) => Promise<{result: T}>

export interface MigratorConfig {
  /** @pgkit/client instance */
  client: Client
  migrationsPath: string
  migrationTableName?: string | string[]
  /**
   * Whether to use `client.transaction(tx => ...)` or `client.connect(cn => ...)` when running up/down migrations
   * @default `transaction`
   */
  connectMethod?: 'transaction' | 'connect'

  /**
   * A function which wraps its callback, can be used to add logging before/after completion. Compatible with [tasuku](https://npmjs.com/package/tasuku), for example.
   * @example Using tasuku
   * ```ts
   * import task from 'tasuku'
   *
   * const migrator = new Migrator(___)
   *
   * migrator.useConfig({task}, async () => {
   *  await migrator.up()
   * })
   * ```
   *
   * @example Using a custom task function
   * ```ts
   * const migrator = new Migrator(___)
   *
   * migrator.useConfig({
   *    task: async (name, fn) => {
   *      migrator.logger.info('Starting', name)
   *      const result = await fn()
   *      migrator.logger.info('Finished', name)
   *      return {result}
   *    },
   *    async () => {
   *      await migrator.up()
   *    })
   * })
   * ```
   */
  task: Task
  logger: Logger

  /** flags passed to [migra](https://npmjs.com/package/@pgkit/migra) - which is called for various introspection tasks */
  defaultMigraOptions?: MigraOptions
}

export interface MigratorConstructorParams extends Omit<MigratorConfig, 'client' | 'task' | 'logger'> {
  client: string | Client
  task?: Task
  logger?: Logger
}
