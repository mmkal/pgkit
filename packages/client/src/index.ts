// codegen:start {preset: barrel}
export * from './client'
export * from './errors'
export * from './naming'
export * from './sql'
export * from './storage'
export * from './type-parsers'
export * from './types'
// codegen:end
