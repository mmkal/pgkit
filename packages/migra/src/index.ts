export {Changes} from './changes'
export {run, type MigraOptions} from './command'
export {Migration} from './migra'
export {Statements, UnsafeMigrationException} from './statements'

export {PostgreSQL} from '@pgkit/schemainspect'
