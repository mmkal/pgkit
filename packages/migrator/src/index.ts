import {createHash} from 'crypto'
import {readFileSync} from 'fs'
import {basename, dirname, join} from 'path'
import * as umzug from 'umzug'
import {sql, DatabaseTransactionConnectionType, DatabasePoolConnectionType, DatabasePoolType} from 'slonik'
import * as path from 'path'
import * as templates from './templates'

interface SlonikMigratorContext {
  parent: DatabasePoolType
  connection: DatabaseTransactionConnectionType
  sql: typeof sql
  commit: () => Promise<void>
}

export class SlonikMigrator extends umzug.Umzug<SlonikMigratorContext> {
  constructor(
    private slonikMigratorOptions: {
      slonik: DatabasePoolType
      migrationsPath: string
      migrationTableName: string | string[]
      logger: umzug.UmzugOptions['logger']
      singleTransaction?: true
    },
  ) {
    super({
      context: () => ({
        parent: slonikMigratorOptions.slonik,
        sql,
        commit: null as never, // commit function is added later by storage setup.
        connection: null as never, // commit function is added later by storage setup.
      }),
      migrations: () => ({
        glob: [this.migrationsGlob(), {cwd: path.resolve(slonikMigratorOptions.migrationsPath)}],
        resolve: params => this.resolver(params),
      }),
      storage: {
        executed: (...args) => this.executedNames(...args),
        logMigration: (...args) => this.logMigration(...args),
        unlogMigration: (...args) => this.unlogMigration(...args),
      },
      logger: slonikMigratorOptions.logger,
      create: {
        template: filepath => this.template(filepath),
        folder: path.resolve(slonikMigratorOptions.migrationsPath),
      },
    })

    if ('mainModule' in slonikMigratorOptions) {
      throw new Error(`Using \`mainModule\` is deprecated. Use \`migrator.runAsCLI()\` instead.`)
    }

    if (!slonikMigratorOptions.migrationTableName) {
      throw new Error(
        `@slonik/migrator: Relying on the default migration table name is deprecated. You should set this explicitly to 'migration' if you've used a prior version of this library.`,
      )
    }

    this.on('beforeCommand', ev => this.setup(ev))
    this.on('afterCommand', ev => this.teardown(ev))
  }

  getCli(options?: umzug.CommandLineParserOptions) {
    return super.getCli({toolDescription: `@slonik/migrator - PostgreSQL migration tool`, ...options})
  }

  async runAsCLI(argv?: string[]) {
    const result = await super.runAsCLI(argv)
    await this.slonikMigratorOptions.slonik.end?.()
    return result
  }

  /** Glob pattern with `migrationsPath` as `cwd`. Could be overridden to support nested directories */
  protected migrationsGlob() {
    return './*.{js,ts,sql}'
  }

  /** Gets a hexadecimal integer to pass to postgres's `select pg_advisory_lock()` function */
  protected advisoryLockId() {
    const hashable = '@slonik/migrator advisory lock:' + JSON.stringify(this.slonikMigratorOptions.migrationTableName)
    const hex = createHash('md5').update(hashable).digest('hex').slice(0, 8)
    return parseInt(hex, 16)
  }

  protected migrationTableNameIdentifier() {
    const table = this.slonikMigratorOptions.migrationTableName
    return sql.identifier(Array.isArray(table) ? table : [table])
  }

  protected template(filepath: string): Array<[string, string]> {
    if (filepath.endsWith('.ts')) {
      return [[filepath, templates.typescript]]
    }
    if (filepath.endsWith('.js')) {
      return [[filepath, templates.javascript]]
    }
    const downPath = path.join(path.dirname(filepath), 'down', path.basename(filepath))
    return [
      [filepath, templates.sqlUp],
      [downPath, templates.sqlDown],
    ]
  }

  protected resolver(
    params: umzug.MigrationParams<SlonikMigratorContext>,
  ): umzug.RunnableMigration<SlonikMigratorContext> {
    if (path.extname(params.name) === '.sql') {
      return {
        name: params.name,
        path: params.path,
        up: async ({path, context}) => {
          await context!.connection.query(rawQuery(readFileSync(path!, 'utf8')))
        },
        down: async ({path, context}) => {
          const downPath = join(dirname(path!), 'down', basename(path!))
          await context!.connection.query(rawQuery(readFileSync(downPath, 'utf8')))
        },
      }
    }
    const {connection: slonik} = params.context
    const migrationModule: {up: Migration; down?: Migration} = require(params.path!)
    return {
      name: params.name,
      path: params.path,
      up: async upParams => migrationModule.up({slonik, sql, ...upParams}),
      down: async downParams => migrationModule.down?.({slonik, sql, ...downParams}),
    }
  }

  protected getOrCreateMigrationsTable(context: SlonikMigratorContext) {
    return context.parent.query(sql`
      create table if not exists ${this.migrationTableNameIdentifier()}(
        name text primary key,
        hash text not null,
        date timestamptz not null default now()
      )
    `)
  }

  protected async setup({context}: {context: SlonikMigratorContext}) {
    let settle!: Function
    const settledPromise = new Promise(resolve => (settle = resolve))

    let ready!: Function
    const readyPromise = new Promise(r => (ready = r))

    const connect = this.slonikMigratorOptions.singleTransaction ? context.parent.transaction : context.parent.connect
    const connectionPromise = connect(async connection => {
      context.connection = connection
      ready()

      await settledPromise
    })

    await readyPromise

    context.commit = () => {
      settle()
      return connectionPromise
    }

    const logger = this.slonikMigratorOptions.logger

    try {
      const timeout = setTimeout(() => logger?.info({message: `Waiting for lock...`} as any), 1000)
      await context.connection.any(context.sql`select pg_advisory_lock(${this.advisoryLockId()})`)
      clearTimeout(timeout)

      await this.getOrCreateMigrationsTable(context)
    } catch (e) {
      await context.commit()
      throw e
    }
  }

  protected async teardown({context}: {context: SlonikMigratorContext}) {
    await context.connection.query(context.sql`select pg_advisory_unlock(${this.advisoryLockId()})`).catch(error => {
      this.slonikMigratorOptions.logger?.error({
        message: `Failed to unlock. This is expected if the lock acquisition timed out.`,
        originalError: error,
      })
    })
    await context.commit()
    context.connection = null as never
  }

  protected hash(name: string) {
    return name
  }

  protected async executedNames({context}: {context: SlonikMigratorContext}) {
    await this.getOrCreateMigrationsTable(context)
    const results = await context.parent
      .any(sql`select name, hash from ${this.migrationTableNameIdentifier()}`)
      .then(migrations =>
        migrations.map(r => {
          const name = r.name as string
          /* istanbul ignore if */
          if (r.hash !== this.hash(name)) {
            this.slonikMigratorOptions.logger?.warn({
              message: `hash in '${this.slonikMigratorOptions.migrationTableName}' table didn't match content on disk.`,
              question: `did you try to change a migration file after it had been run?`,
              migration: r.name,
              dbHash: r.hash,
              diskHash: this.hash(name),
            })
          }
          return name
        }),
      )

    return results
  }

  protected async logMigration({name, context}: {name: string; context: SlonikMigratorContext}) {
    await context.connection.query(sql`
      insert into ${this.migrationTableNameIdentifier()}(name, hash)
      values (${name}, ${this.hash(name)})
    `)
  }

  protected async unlogMigration({name, context}: {name: string; context: SlonikMigratorContext}) {
    await context.connection.query(sql`
      delete from ${this.migrationTableNameIdentifier()}
      where name = ${name}
    `)
  }
}

export type Migration = (
  params: umzug.MigrationParams<SlonikMigratorContext> & {
    /** @deprecated use `context.connection` */
    slonik: DatabaseTransactionConnectionType
    /** @deprecated use `context.sql` */
    sql: typeof sql
  },
) => Promise<unknown>

/**
 * Should either be a `DatabasePoolType` or `DatabasePoolConnectionType`. If it's a `DatabasePoolType` with an `.end()`
 * method, the `.end()` method will be called after running the migrator as a CLI.
 */
export interface SlonikConnectionType extends DatabasePoolConnectionType {
  end?: () => Promise<void>
}

export interface SlonikMigratorOptions {
  /**
   * Slonik instance for running migrations. You can import this from the same place as your main application,
   * or import another slonik instance with different permissions, security settings etc.
   */
  slonik: DatabasePoolType
  /**
   * Path to folder that will contain migration files.
   */
  migrationsPath: string
  /**
   * REQUIRED table name. @slonik/migrator will manage this table for you, but you have to tell it the name
   * Note: prior to version 0.6.0 this had a default value of "migration", so if you're upgrading from that
   * version, you should name it that!
   */
  migrationTableName: string | string[]
  /**
   * Logger with `info`, `warn`, `error` and `debug` methods - set explicitly to `undefined` to disable logging
   */
  logger: umzug.UmzugOptions['logger']
}

/**
 * More reliable than slonik-sql-tag-raw: https://github.com/gajus/slonik-sql-tag-raw/issues/6
 * But doesn't sanitise any inputs, so shouldn't be used with templates
 */
const rawQuery = (query: string): ReturnType<typeof sql> => ({
  type: 'SLONIK_TOKEN_SQL',
  sql: query,
  values: [],
})

/**
 * Narrowing of @see umzug.UmzugOptions where the migrations input type specifically, uses `glob`
 */
export type SlonikUmzugOptions = umzug.UmzugOptions<SlonikMigratorContext> & {
  migrations: umzug.GlobInputMigrations<SlonikMigratorContext>
}

/**
 * @deprecated use `new SlonikMigrator(...)` which takes the same options.
 *
 * Note: `mainModule` is not passed into `new SlonikMigrator(...)`. To get the same functionality, use `.runAsCLI()`
 *
 * @example
 * ```
 * const migrator = new SlonikMigrator(...)
 *
 * if (require.main === module) {
 *  migrator.runAsCLI()
 * }
 * ```
 */
export const setupSlonikMigrator = (
  options: SlonikMigratorOptions & {
    /**
     * @deprecated Use `.runAsCLI()`, e.g. `if (require.main === module) migrator.runAsCLI()`
     *
     * ~OPTIONAL "module" value. If you set `mainModule: module` in a nodejs script, that script will become a
     * runnable CLI when invoked directly, but the migrator object can still be imported as normal if a different
     * entrypoint is used.~
     */
    mainModule?: NodeModule
    reasonForUsingDeprecatedAPI: 'Back-compat' | 'Testing' | `Life's too short` | 'Other'
  },
) => {
  console.warn(
    `@slonik/migrator: Use of ${setupSlonikMigrator.name} is deprecated. Use \`new SlonikMigrator(...)\` which takes the same options instead`,
  )
  const defaultMigrationTableName = () => {
    console.warn(
      `Relying on the default migration table name is deprecated. You should set this explicitly to 'migration'`,
    )
    return 'migration'
  }
  const migrator = new SlonikMigrator({
    slonik: options.slonik,
    migrationsPath: options.migrationsPath,
    migrationTableName: options.migrationTableName || defaultMigrationTableName(),
    logger: options.logger,
  })
  if (options.mainModule === require.main) {
    console.warn(`Using \`mainModule\` is deprecated. Use \`migrator.runAsCLI()\` instead.`)
    migrator.runAsCLI()
  }
  return migrator
}
