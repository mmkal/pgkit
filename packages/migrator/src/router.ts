import {sql} from '@pgkit/client'
import * as trpcServer from '@trpc/server'
import {readFile} from 'fs/promises'
import z from 'zod'
import {Migrator} from './migrator'
import {Confirm} from './types'

export interface MigratorRouterContext {
  migrator: Migrator
  confirm: Confirm
}

/** Based on `cleye`'s `HelpOptions` */
export interface MigratorRouterMeta {
  /** Description of the script or command to display in `--help` output. */
  description?: string
  /** Usage code examples to display in `--help` output. */
  usage?: false | string | string[]
  /** Example code snippets to display in `--help` output. */
  examples?: string | string[]
}

/** Helper type that IMO should exist in trpc. Basically a type which a trpc-procedure with context of type `Ctx` will satisfy */
export type TRPCProcedureLike<Ctx> = {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  mutation: (resolver: (opts: {ctx: Ctx; input: any}) => any) => any
}

/** Parameters needed for a helper function returning a router, which can be used as a sub-router for another. */
export interface TRPCHelperParams<Ctx> {
  /** The `trpc.procedure` function. A middleware can be used to align context types. */
  procedure: TRPCProcedureLike<Ctx>
}

export const migratorTrpc = trpcServer.initTRPC
  .context<MigratorRouterContext>()
  .meta<MigratorRouterMeta>()
  .create() satisfies TRPCHelperParams<MigratorRouterContext> // use `satisfies` to make sure the `TRPCProcedureLike` type helper is correct

/**
 * Get a router with procedures needed to manage migrations.
 *
 * @param procedure - TRPC procedure builder for the router this will be attached to. Must have the correct context type
 *
 * @example
 * import {createMigratorRouter, Migrator} from '@pgkit/migrator'
 * import {initTRPC} from '@trpc/server'
 *
 * const t = initTRPC.context<YourAppContext>().create()
 *
 * export const yourAppRouter = t.router({
 *   yourProcedeure: t.procedure.query(async () => 'Hello, world!'),
 *   migrations: getMigrationsRouter(),
 * })
 *
 * function getMigrationsRouter() {
 *   return createMigratorRouter(
 *     t.procedure.use(async ({ctx, next}) => {
 *       return next({
 *         ctx: {
 *           migrator: new Migrator(___),
 *           confirm: async (sql: string) => {
 *             return ctx.whitelistedSql.includes(sql)
 *           },
 *         }
 *       })
 *     })
 *   )
 * }
 */
export const createMigratorRouter = (procedure: TRPCProcedureLike<MigratorRouterContext> = migratorTrpc.procedure) => {
  // Take advantage of trpc being overly lax about merging routers with different contexts: https://github.com/trpc/trpc/issues/4306
  // We have used the `TRPCProcedureLike` type to ensure that the context is correct for the procedure builder, and trpc will merge the routers without checking the context
  // This means any router with a different context type can use this helper to creater a migrations sub-router, by just defining a middleware that sets the correct context
  const trpc = {router: migratorTrpc.router, procedure: procedure} as typeof migratorTrpc

  const router = trpc.router({
    up: trpc.procedure
      .meta({description: 'Apply pending migrations'})
      .input(
        z
          .union([
            z.object({}).strict(), //
            z.object({to: z.string().describe('Only apply migrations up to this one')}),
            z.object({step: z.number().int().positive().describe('Apply this many migrations')}),
          ])
          .optional(),
      )
      .mutation(async ({input, ctx}) => {
        return ctx.migrator.up(input)
      }),
    create: trpc.procedure
      .meta({description: 'Create a new migration file'})
      .input(
        z.object({
          content: z
            .string()
            .optional()
            .describe(
              'SQL content of the migration. If not specified, content will be generated based on the calculated diff between the existing migrations and the current database state.',
            ),
          name: z
            .string()
            .optional()
            .describe(
              'Name of the migration file. If not specified, a name will be generated based on the content of the migraiton',
            ),
        }),
      )
      .mutation(async ({input, ctx}) => {
        return ctx.migrator.create(input)
      }),
    list: trpc.procedure
      .meta({
        description: 'List migrations, along with their status, file path and content',
      })
      .input(
        z.object({
          query: z
            .string()
            .optional()
            .describe('Search query - migrations with names containing this string will be returned'),
          status: z.enum(['pending', 'executed']).optional().describe('Filter by status'),
          result: z
            .enum(['first', 'last', 'one', 'maybeOne', 'all'])
            .default('all')
            .describe('Which result(s) to return'),
          output: z
            .enum(['name', 'path', 'content', 'object'])
            .default('object')
            .describe('Result properties to return'),
        }),
      )
      .query(async ({input, ctx}) => {
        const list = await ctx.migrator.list()
        const results = list
          .filter(m => {
            return m.name.includes(input.query || '') && m.status === (input.status || m.status)
          })
          .map(m => {
            if (input.output === 'name') return m.name
            if (input.output === 'path') return m.path
            if (input.output === 'content') return m.content
            return m
          })

        if (input.result === 'all') return results

        if (input.result === 'one' && results.length !== 1) {
          throw new Error(
            `Expected exactly one migration matching query ${JSON.stringify(input.query)}, found ${results.length}`,
            {
              cause: {results},
            },
          )
        }

        if (input.result === 'maybeOne' && results.length !== 1) return undefined
        if (input.result === 'first') return results[0]
        if (input.result === 'last') return results.at(-1)
      }),
    latest: trpc.procedure
      .meta({
        description: 'Get the latest migration',
      })
      .input(
        z.object({
          skipCheck: z.boolean().optional().describe('Skip checking that migrations are in a valid state'),
        }),
      )
      .query(async ({input, ctx}) => {
        return ctx.migrator.latest(input)
      }),
    check: trpc.procedure
      .meta({description: 'Verify that your database is in an expected state, matching your migrations'})
      .mutation(async ({ctx}) => {
        return ctx.migrator.check()
      }),
    repair: trpc.procedure
      .meta({
        description:
          'If your migrations are not in a valid state, this will calculate the diff required to move your database to a valid state, and apply it',
      })
      .mutation(async ({ctx}) => {
        return ctx.migrator.repair({confirm: ctx.confirm})
      }),
    goto: trpc.procedure
      .meta({
        description:
          'Go "back" to a specific migration. This will calculate the diff required to get to the target migration, then apply it',
      })
      .input(
        z.object({
          name: z.string().describe('Name of the migration to go to. Use "list" to see available migrations.'),
        }),
      )
      .mutation(async ({input, ctx}) => {
        return ctx.migrator.goto({name: input.name, confirm: ctx.confirm})
      }),
    baseline: trpc.procedure
      .meta({
        description:
          'Baseline the database at the specified migration. This forcibly edits the migrations table to mark all migrations up to this point as executed. Useful for introducing the migrator to an existing database.',
      })
      .input(
        z.object({
          to: z.string().describe('Name of the migration to baseline to. Use `list` to see available migrations.'),
          purgeDisk: z.boolean().default(false).describe('Delete files subsequent to the specified migration'),
        }),
      )
      .mutation(async ({input, ctx}) => {
        return ctx.migrator.baseline({...input, to: input.to, confirm: ctx.confirm})
      }),
    rebase: trpc.procedure
      .meta({
        description:
          'Rebase the migrations from the specified migration. This deletes all migration files after this point, and replaces them with a squashed migration based on the calculated diff required to reach the current database state.',
      })
      .input(
        z.object({
          from: z
            .string()
            .describe(
              'Name of the migration to rebase from. This migration will remain, all subsequent ones will be replaced with a squashed migration. Use `list` to see available migrations.',
            ),
        }),
      )
      .mutation(async ({input, ctx}) => {
        return ctx.migrator.rebase({...input, confirm: ctx.confirm})
      }),
    definitions: trpc.router({
      filepath: trpc.procedure
        .meta({
          description: 'Get the path to the definitions file',
        })
        .query(async ({ctx}) => {
          return {
            path: ctx.migrator.definitionsFile,
            content: await readFile(ctx.migrator.definitionsFile, 'utf8').catch(() => null),
          }
        }),
      updateDb: trpc.procedure
        .meta({description: 'Update the database from the definitions file'})
        .mutation(async ({ctx}) => {
          return ctx.migrator.updateDbToMatchDefinitions({confirm: ctx.confirm})
        }),
      updateFile: trpc.procedure
        .meta({description: 'Update the definitions file from the database'})
        .mutation(async ({ctx}) => {
          return ctx.migrator.updateDefinitionsToMatchDb({confirm: ctx.confirm})
        }),
    }),
    unlock: trpc.procedure
      .meta({
        description:
          'Release the advisory lock for this migrator on the database. This is useful if the migrator is stuck due to a previous crash',
      })
      .mutation(async ({ctx}) => {
        return ctx.migrator.unlock({confirm: ctx.confirm})
      }),
    wipe: trpc.procedure
      .meta({
        description: 'Wipe the database - remove all tables, views etc.',
      })
      .mutation(async ({ctx}) => {
        return ctx.migrator.wipe({confirm: ctx.confirm})
      }),
    sql: trpc.procedure
      .meta({
        description:
          'Query the database. Not strictly related to migrations, but can be used for debugging. Use with caution!',
      })
      .input(
        z.tuple([
          z.string(),
          z.object({
            singlequote: z
              .string()
              .describe("Character to use in place of ' - use to avoid having to do bash quote-escaping")
              .optional(),
            doublequote: z
              .string()
              .describe('Character to use in place of " - use to avoid having to do bash quote-escaping')
              .optional(),
            method: z
              .enum(['any', 'many', 'one', 'maybeOne', 'query', 'anyFirst', 'oneFirst', 'maybeOneFirst'])
              .default('any'),
          }),
        ]),
      )
      .mutation(async ({input: [query, options], ctx}) => {
        if (options.singlequote) query = query.replaceAll(options.singlequote, `'`)
        if (options.doublequote) query = query.replaceAll(options.doublequote, `"`)

        return ctx.migrator.client[options.method](sql.raw(query))
      }),
  })

  return router
}
