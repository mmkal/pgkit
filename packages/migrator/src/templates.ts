export const sqlUp = `raise 'up migration not implemented'
`

export const sqlDown = `raise 'down migration not implemented'
`

export const typescript = `
import {Migration} from '@slonik/migrator'

export const up: Migration = async ({slonik, sql}) => {
  await slonik.query(sql\`raise 'up migration not implemented'\`)
}

export const down: Migration = async ({slonik, sql}) => {
  await slonik.query(sql\`raise 'down migration not implemented'\`)
}
`.trimLeft()

export const javascript = `
/** @type {import('@sloink/migrator').Migration} */
exports.up = async ({slonik, sql}) => {
  await slonik.query(sql\`raise 'up migration not implemented'\`)
}

/** @type {import('@sloink/migrator').Migration} */
exports.down = async ({slonik, sql}) => {
  await slonik.query(sql\`raise 'down migration not implemented'\`)
}
`.trimLeft()
