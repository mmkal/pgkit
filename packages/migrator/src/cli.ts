import * as prompt from '@inquirer/prompts'
import * as childProcess from 'child_process'
import * as fs from 'fs'
import * as os from 'os'
import * as path from 'path'
import * as colors from 'picocolors'
import {createCli} from 'trpc-cli'
import {Migrator} from './migrator'
import {createMigratorRouter} from './router'

export const getMigratorFromEnv = () => {
  return new Migrator({
    client: process.env.PGKIT_CONNECTION_STRING || 'postgresql://postgres:postgres@localhost:5432/postgres',
    migrationsPath: process.env.PGKIT_MIGRATIONS_PATH || path.join(process.cwd(), 'migrations'),
    migrationTableName: process.env.PGKIT_MIGRATIONS_TABLE_NAME || 'pgkit_migrations',
  })
}

export const createMigratorCli = (migrator: Migrator) => {
  return createCli({
    router: createMigratorRouter(),
    context: {migrator, confirm},
  })
}

export type MigratorCLI = ReturnType<typeof createMigratorCli>

export const confirm = async (input: string, options?: {readonly?: boolean}): Promise<string | null> => {
  let currentInput = input

  while (currentInput.trim()) {
    const message = `${colors.underline('Please confirm you want to run the following')}:\n\n${currentInput}`
    const editors = [
      ...new Set([process.env.VISUAL, process.env.EDITOR, 'cursor', 'code', 'vi', 'emacs', 'nano', 'notepad']),
    ].filter(name => {
      return !name?.match(/\W/) && childProcess.execSync(`sh -c 'which ${name} || echo ""'`).toString().trim()
    })
    const choice = await prompt.select({
      message,
      choices: [
        {name: 'Yes', value: 'yes'},
        {name: 'No', value: 'no'},
        ...editors.map(name => ({
          name: `Edit with ${name}`,
          disabled: options?.readonly,
          description: `Save and close the file after editing. You'll be asked again if you want to run the edit migration afterwards.`,
          value: `edit-${name}`,
        })),
      ],
    })

    if (choice === 'yes') return currentInput
    if (choice === 'no') return null

    currentInput = await editTempFile(currentInput, choice.replace('edit-', ''))
  }

  return null
}

async function editTempFile(input: string, editor = process.env.EDITOR || 'vi'): Promise<string> {
  const tempFile = path.join(os.tmpdir(), `pgkit-migrator`, `changes-${Date.now()}.sql`)
  fs.mkdirSync(path.dirname(tempFile), {recursive: true})
  fs.writeFileSync(tempFile, input.trim() + '\n')

  try {
    await new Promise<void>((resolve, reject) => {
      const child = childProcess.spawn(editor, [tempFile], {stdio: 'inherit'})
      child.on('exit', code => {
        if (code === 0) resolve()
        else reject(new Error(`Editor exited with code ${code}`))
      })
    })

    // Wait for user input to ensure the file has been saved
    await prompt.input({message: 'Press Enter when you have finished editing'})

    const editedContent = fs.readFileSync(tempFile, 'utf8')
    return editedContent.trim()
  } finally {
    fs.unlinkSync(tempFile)
  }
}
