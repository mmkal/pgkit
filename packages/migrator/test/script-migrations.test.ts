import {join} from 'path'
import {range} from 'lodash'
import {sql} from 'slonik'
import * as dedent from 'dedent'
import {fsSyncer} from 'fs-syncer'
import {SlonikMigrator} from '../src'
import {getPoolHelper} from './pool-helper'

const helper = getPoolHelper({__filename})

const millisPerDay = 1000 * 60 * 60 * 24
const fakeDates = range(0, 100).map(days => new Date(new Date('2000').getTime() + days * millisPerDay).toISOString())

const toISOSpy = jest.spyOn(Date.prototype, 'toISOString')
toISOSpy.mockImplementation(() => fakeDates[toISOSpy.mock.calls.length - 1])

describe('run sql, js and ts migrations', () => {
  const migrationsPath = join(__dirname, 'generated/run/migrations')

  const mockLogger = jest.fn()
  const log = mockLogger

  const syncer = fsSyncer(migrationsPath, {
    '01.one.sql': 'create table migration_test_1(id int)',
    '02.two.sql': 'create table migration_test_2(id int)',
    '03.three.js': dedent`
      module.exports.up = ({context: {connection, sql}}) => connection.query(sql\`create table migration_test_3(id int)\`)
      module.exports.down = ({context: {connection, sql}}) => connection.query(sql\`drop table migration_test_3\`)
    `,
    '04.four.ts': dedent`
      import {Migration} from '../../../..'

      export const up: Migration = ({context: {connection, sql}}) => connection.query(sql\`create table migration_test_4(id int)\`)
      export const down: Migration = ({context: {connection, sql}}) => connection.query(sql\`drop table migration_test_4\`)
    `,
    down: {
      '01.one.sql': 'drop table migration_test_1',
      '02.two.sql': 'drop table migration_test_2',
    },
  })

  beforeEach(async () => {
    syncer.sync()
  })

  const migrationTables = () =>
    helper.pool.anyFirst(sql`
      select tablename
      from pg_catalog.pg_tables
      where tablename like 'migration_test_%'
      and schemaname = ${helper.schemaName}
      order by tablename
    `)

  test('up and down', async () => {
    const migrator = new SlonikMigrator({
      slonik: helper.pool,
      migrationsPath,
      migrationTableName: 'migration_meta_1',
      logger: {
        debug: log,
        info: log,
        warn: log,
        error: log,
      },
    })

    expect(await migrationTables()).toEqual([])

    const executed = () => migrator.executed().then(list => list.map(x => x.name))
    const pending = () => migrator.pending().then(list => list.map(x => x.name))

    log(`getting pending migrations before initial 'up'`)

    const allMigrations = await pending()
    expect(allMigrations).toMatchInlineSnapshot(`
      Array [
        "01.one.sql",
        "02.two.sql",
        "03.three.js",
        "04.four.ts",
      ]
    `)

    await migrator.up()

    const allTables = await migrationTables()
    expect(allTables).toMatchInlineSnapshot(`
      Array [
        "migration_test_1",
        "migration_test_2",
        "migration_test_3",
        "migration_test_4",
      ]
    `)

    expect(await executed()).toEqual(allMigrations)
    expect(await pending()).toEqual([])

    await migrator.down({to: 0})
    expect(await executed()).toEqual([])

    await migrator.up({to: '03.three.js'})
    expect(await executed()).toMatchInlineSnapshot(`
      Array [
        "01.one.sql",
        "02.two.sql",
        "03.three.js",
      ]
    `)

    expect(
      mockLogger.mock.calls.map(msg => {
        const json = JSON.stringify(msg)
        return JSON.parse(json.replace(/\d\.\d+/g, '0.001'))
      }),
    ).toMatchSnapshot()
  })
})
