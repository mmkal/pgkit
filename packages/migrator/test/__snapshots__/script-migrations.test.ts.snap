// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`run sql, js and ts migrations up and down 4`] = `
Array [
  Array [
    "getting pending migrations before initial 'up'",
  ],
  Array [
    Object {
      "event": "migrating",
      "name": "01.one.sql",
    },
  ],
  Array [
    Object {
      "durationSeconds": 0.001,
      "event": "migrated",
      "name": "01.one.sql",
    },
  ],
  Array [
    Object {
      "event": "migrating",
      "name": "02.two.sql",
    },
  ],
  Array [
    Object {
      "durationSeconds": 0.001,
      "event": "migrated",
      "name": "02.two.sql",
    },
  ],
  Array [
    Object {
      "event": "migrating",
      "name": "03.three.js",
    },
  ],
  Array [
    Object {
      "durationSeconds": 0.001,
      "event": "migrated",
      "name": "03.three.js",
    },
  ],
  Array [
    Object {
      "event": "migrating",
      "name": "04.four.ts",
    },
  ],
  Array [
    Object {
      "durationSeconds": 0.001,
      "event": "migrated",
      "name": "04.four.ts",
    },
  ],
  Array [
    Object {
      "event": "reverting",
      "name": "04.four.ts",
    },
  ],
  Array [
    Object {
      "durationSeconds": 0.001,
      "event": "reverted",
      "name": "04.four.ts",
    },
  ],
  Array [
    Object {
      "event": "reverting",
      "name": "03.three.js",
    },
  ],
  Array [
    Object {
      "durationSeconds": 0.001,
      "event": "reverted",
      "name": "03.three.js",
    },
  ],
  Array [
    Object {
      "event": "reverting",
      "name": "02.two.sql",
    },
  ],
  Array [
    Object {
      "durationSeconds": 0.001,
      "event": "reverted",
      "name": "02.two.sql",
    },
  ],
  Array [
    Object {
      "event": "reverting",
      "name": "01.one.sql",
    },
  ],
  Array [
    Object {
      "durationSeconds": 0.001,
      "event": "reverted",
      "name": "01.one.sql",
    },
  ],
  Array [
    Object {
      "event": "migrating",
      "name": "01.one.sql",
    },
  ],
  Array [
    Object {
      "durationSeconds": 0.001,
      "event": "migrated",
      "name": "01.one.sql",
    },
  ],
  Array [
    Object {
      "event": "migrating",
      "name": "02.two.sql",
    },
  ],
  Array [
    Object {
      "durationSeconds": 0.001,
      "event": "migrated",
      "name": "02.two.sql",
    },
  ],
  Array [
    Object {
      "event": "migrating",
      "name": "03.three.js",
    },
  ],
  Array [
    Object {
      "durationSeconds": 0.001,
      "event": "migrated",
      "name": "03.three.js",
    },
  ],
]
`;
